bitcoin/script/miniscript.h

```diff
@@ -1014,224 +1014,384 @@ bool DecomposeScript(const CScript& script, std::vector<std::pair<opcodetype, st
+                to_parse.emplace_back(DecodeContext::Expression, -1, -1);
+                to_parse.emplace_back(DecodeContext::WExpression, -1, -1);
             }
-            if (last > in) {
-                auto sub3 = DecodeSingle<Key>(in, last, ctx);
-                if (!sub3) return {};
-                return MakeNodeRef<Key>(NodeType::OR_C, Vector(std::move(sub3), std::move(sub1)));
+            else {
+                // Unrecognised expression
+                return {};
             }
-            return {};
+            break;
         }
-        if (have_else && !negated) {
-            return MakeNodeRef<Key>(NodeType::OR_I, Vector(std::move(sub2), std::move(sub1)));
+        case DecodeContext::WExpression: {
+            // a: wrapper
+            if (in >= last) return {};
+            if (in[0].first == OP_FROMALTSTACK) {
+                ++in;
+                to_parse.emplace_back(DecodeContext::Alt, -1, -1);
+            } else {
+                to_parse.emplace_back(DecodeContext::Swap, -1, -1);
+            }
+            to_parse.emplace_back(DecodeContext::MaybeAndV, -1, -1);
+            to_parse.emplace_back(DecodeContext::Expression, -1, -1);
+            break;
         }
-        return {};
-    }
-    keys.clear();
-    if (last - in >= 3 && in[0].first == OP_CHECKMULTISIG) {
-        int64_t n;
-        if (!ParseScriptNumber(in[1], n)) return {};
-        if (last - in < 3 + n) return {};
-        if (n < 1 || n > 20) return {};
-        for (int i = 0; i < n; ++i) {
-            Key key;
-            if (in[2 + i].second.size() != 33) return {};
-            if (!ctx.FromPKBytes(in[2 + i].second.begin(), in[2 + i].second.end(), key)) return {};
-            keys.push_back(std::move(key));
+        case DecodeContext::MaybeAndV: {
+            // If we reach a potential AND_V top-level, check if the next part of the script could be another AND_V child
+            if (in < last && in[0].first != OP_IF && in[0].first != OP_ELSE && in[0].first != OP_NOTIF && in[0].first != OP_TOALTSTACK && in[0].first != OP_SWAP) {
+                to_parse.emplace_back(DecodeContext::AndV, -1, -1);
+                to_parse.emplace_back(DecodeContext::Expression, -1, -1);
+            }
+            break;
         }
-        if (!ParseScriptNumber(in[2 + n], k)) return {};
-        if (k < 1 || k > n) return {};
-        in += 3 + n;
-        std::reverse(keys.begin(), keys.end());
-        return MakeNodeRef<Key>(NodeType::MULTI, std::move(keys), k);
-    }
-    subs.clear();
-    if (last - in >= 3 && in[0].first == OP_EQUAL && ParseScriptNumber(in[1], k)) {
-        if (k < 1) {
-            return {};
+        case DecodeContext::Swap: {
+            if (in >= last || in[0].first != OP_SWAP) return {};
+            ++in;
+            NodeRef<Key> node = MakeNodeRef<Key>(NodeType::WRAP_S, Vector(std::move(constructed.back())));
+            constructed.pop_back();
```

Comment:
In a bunch of places you have these `constructed.pop_back(); contructed.push_back(std::move(node));` lines; it's simpler and more efficient to use `constructed.back() = std::move(node);`.

`contructed.back() = MakeNodeRef<Key>(...,...std::move(constructed.back())...)` would also work, FWIW, but that may be a bit too dense. Up to you.